### PR TITLE
Fix issue 23844: Support ranges with immutable fields (like `only` with `immutable struct`) in `chain`.

### PR DESCRIPTION
**Pull request by FeepingCreature**
Fixes issue 23844.

`only` is a range that may be *mutable*, but not *assignable*. `chain` falls over here because it assumes it can make a struct with ranges, and reassign them with new values, which isn't necessarily the case even if the ranges are not `const`. Solved by creating a separate tuple of `Rebindable` ranges for this case.

### CODE REVIEW

**Comment by FeepingCreature** [comment]
Ping~

File: std/range/package.d [hunk 1]
```diff
@@ -1404,7 +1418,10 @@ if (Ranges.length > 0 &&
                         assert(0, "Internal library error. Please report it.");
                     }
 
-                    return result;
+                    static if (bidirectional)
```
Comment (atilaneves) [inline std/range/package.d:1421]: How is this related?

Comment (FeepingCreature) [reply std/range/package.d:1421]: Further down, we used to mutate `result.backIndex` in the `static if (bidirectional)` branch. As `result` no longer exists, we have to provide a variable for that branch to mutate.

**Comment by FeepingCreature** [comment]
Ping :)

**Comment by dkorpel** [comment]
@atilaneves

**Comment by FeepingCreature** [comment]
And again.

**Comment by FeepingCreature** [comment]
To be clear, this PR is *not* blocked by #8735. #8735 will simplify use of `Rebindable`, but it's entirely viable without it. The big thing (struct support) is already merged.

**Review by dkorpel — APPROVED** [review]
